### PR TITLE
fix(node:http) fix node:http chunked encoding on server and add chunked encoding support on the client

### PR DESCRIPTION
**Pull request by cirospaciari**
### What does this PR do?
This depends on or replaces: https://github.com/oven-sh/bun/pull/15576

Fix: https://github.com/oven-sh/bun/issues/11621
Fix: https://github.com/oven-sh/bun/issues/8638
Fix: https://github.com/oven-sh/bun/issues/11938
Fix: https://github.com/oven-sh/bun/issues/7890
And of course the same ones as #15576:

Fix: https://github.com/oven-sh/bun/issues/6017
Fix: https://github.com/oven-sh/bun/issues/7206
Fix: https://github.com/oven-sh/bun/issues/15022

Improve on: https://github.com/oven-sh/bun/issues/9100

~~Probably https://github.com/oven-sh/bun/pull/13778 should be merged after this if tests pass.~~

- [ ] Documentation or TypeScript types (it's okay to leave the rest blank in this case)
- [x] Code changes

### How did you verify your code works?
Tests

### CODE REVIEW

**Comment by robobun** [comment]
@cirospaciari, your commit e3fa8aea7f18953a00c369d7448cd64f170b3065 is building: [`#7276`](https://buildkite.com/bun/bun/builds/7276)

File: test/js/web/fetch/fetch-leak-test-fixture-5.js [hunk 1]
```diff
@@ -66,6 +66,19 @@ function getBody() {
...
       body = getURLSearchParams();
       break;
+    case "iterator":
+      body = async function* iter() {
+        yield (cachedBody ??= getString());
+      };
+      break;
+    case "stream":
+      body = new ReadableStream({
+        start(c) {
```
Comment (Jarred-Sumner) [inline test/js/web/fetch/fetch-leak-test-fixture-5.js]: This should use pull instead of start and have a slight delay in pull. We may detect if the ReadableStream has already ended with a single chunk and do something else

File: test/js/node/http/node-http.test.ts [hunk 2]
```diff
@@ -2435,3 +2435,118 @@ it("should work when sending https.request with agent:false", async () => {
   await promise;
 });
 
+it("client should use chunked encoded if more than one write is called", async () => {
+  using server = Bun.serve({
```
Comment (Jarred-Sumner) [inline test/js/node/http/node-http.test.ts:2443]: Can you write this test in a way that runs in Node too? This file has had many tests which did not actually run in Node

File: src/js/node/http.ts [hunk 3]
```diff
@@ -838,12 +841,15 @@ IncomingMessage.prototype = {
...
     }
 
-    const contentLength = this.headers["content-length"];
-    const length = contentLength ? parseInt(contentLength, 10) : 0;
-    if (length === 0) {
-      this[noBodySymbol] = true;
-      callback();
-      return;
+    const encoding = this.headers["transfer-encoding"];
+    if (encoding?.indexOf("chunked") === -1) {
```
Comment (Jarred-Sumner) [inline src/js/node/http.ts]:
```suggestion
    if (encoding?.indexOf?.("chunked") === -1) {
```

File: src/js/node/http.ts [hunk 4]
```diff
@@ -777,6 +777,9 @@ function requestHasNoBody(method, req) {
   if ("GET" === method || "HEAD" === method || "TRACE" === method || "CONNECT" === method || "OPTIONS" === method)
     return true;
   const headers = req?.headers;
+  const encoding = headers?.["transfer-encoding"];
+  if (encoding?.indexOf("chunked") !== -1) return false;
```
Comment (Jarred-Sumner) [inline src/js/node/http.ts]:
```suggestion
  if (encoding?.indexOf?.("chunked") !== -1) return false;
```

File: test/js/node/http/node-http.test.ts (hunk 2 above)
Comment (cirospaciari) [reply test/js/node/http/node-http.test.ts:2443]: You mean using node:http server instead of Bun?

Comment (cirospaciari) [reply test/js/node/http/node-http.test.ts:2443]: done

**Comment by Jarred-Sumner** [comment]
Are there node tests we can copy?

File: src/bun.js/webcore/body.zig [hunk 5]
```diff
@@ -158,6 +158,18 @@ pub const Body = struct {
 
             return false;
         }
+
+        pub fn isDisturbed2(this: *const PendingValue, globalObject: *JSC.JSGlobalObject) bool {
```
Comment (Jarred-Sumner) [inline src/bun.js/webcore/body.zig:162]: why can't we use isDisturbed?

File: src/bun.js/event_loop.zig [hunk 6]
```diff
@@ -743,7 +743,7 @@ pub const DeferredTaskQueue = struct {
             };
 
             if (!this.map.values()[i](key)) {
-                this.map.swapRemoveAt(i);
+                _ = this.map.swapRemove(key);
```
Comment (Jarred-Sumner) [inline src/bun.js/event_loop.zig]: Was this change intentional?

File: src/bun.js/webcore/streams.zig [hunk 7]
```diff
@@ -2600,7 +2602,266 @@ pub fn HTTPServerWritable(comptime ssl: bool) type {
...
+    }
+
+    pub fn flush(this: *@This()) JSC.Maybe(void) {
+        _ = this.internalFlush() catch 0;
+        return .{ .result = {} };
+    }
+    pub fn flushFromJS(this: *@This(), globalThis: *JSGlobalObject, _: bool) JSC.Maybe(JSValue) {
+        return .{ .result = JSC.JSPromise.resolvedPromiseValue(globalThis, JSValue.jsNumber(this.internalFlush() catch 0)) };
+    }
+    pub fn destroy(this: *@This()) void {
```
Comment (Jarred-Sumner) [inline src/bun.js/webcore/streams.zig]: use `pub usingnamespace bun.New`

We want that memory leak tracking

File: src/bun.js/event_loop.zig (hunk 6 above)
Comment (cirospaciari) [reply src/bun.js/event_loop.zig]: Yes, was crashing without it

File: src/bun.js/webcore/body.zig (hunk 5 above)
Comment (cirospaciari) [reply src/bun.js/webcore/body.zig:162]: Because can or not be an Request instance or an unknow instance, we only care about the readable stream. (We still use isDisturbed when we know that is a Request instance for example)

File: src/bun.js/event_loop.zig (hunk 6 above)
Comment (cirospaciari) [reply src/bun.js/event_loop.zig]: fixed
